### PR TITLE
Check-ins backend

### LINKED ISSUE #6662: Reading Log Check-Ins
This feature is required for Yearly Reading goals:

When someone clicks currently reading or already read, present options underneath the button to:
- add/update started date (today by default)
- add a check-in
- add a completed date (today by default)
- if "currently reading" end date shouldn't appear
- if a started date is set, we could show "you started reading on [date]" with e.g. a dotted underline which lets patron update date. (this is meant to be illustrative design-wise, not prescriptive).

Constraints: Add multiple, manually editable "Started-Reading" & "Finished-Reading" Dates

### PR DESCRIPTION
**Pull request by jimchamp**
Closes #6662

Adds new DB table for tracking check-in events, and creates the persistence layer for the new table.  Creates two check-in event handlers: one for creating events (`/check-ins/{work_olid}`, and one for updating existing events (`/check-ins/{username}`.  Adds a __temporary__ UI from which to create new check-in events.  New UI is only available to `admin`s.

### Technical

For convenience, the new UI has inline styling and Javascript.  I figure that is okay, as we will not be surfacing this __temporary__ UI to patrons.  The purpose of the UI is for admins to quickly add new events to their dev instances, and to give an example of how to construct a valid payload client-side.  The test UI can be viewed at `/check-ins/{work_olid}` (work OLID need not be valid).

Any updates to the `data` field will overwrite the previous value.  I'm not quite sure what goes in `data` yet, so I don't know if it makes sense to overwrite `data` or to update the object with the new keys.

The endpoints don't really make sense now.  The event creation endpoint has the work OLID in the path, but these events are edition-level.  The endpoint for updating events has a `username` in the path, but we probably want to get the username from `accounts.get_current_user()` to more easily ensure that the request was POSTed by the owner the event (meaning one should only be able to update one's own events).

A new `@authorized_for` decorator has been created in support of this PR.  It does the following:
1. Checks to see if the patron is logged in. If not, a `401` error is returned.
2. Checks to see if the patron is a member of any of the given usergroups.  If not, a `403` error is returned.

### CODE REVIEW

File: openlibrary/plugins/upstream/checkins.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        if data['event_type'] not in BookshelvesEvents.EVENT_TYPES:
+            return False
+        return True
+
+
+class patron_check_ins(delegate.page):
+    path = r'/check-ins/people/([^/]+)'
+    encoding = 'json'
+
+    def POST(self, username):
```
Comment (jimchamp) [inline openlibrary/plugins/upstream/checkins.py:81]:
```suggestion
    @authorized_for('/usergroup/admin')
    def POST(self, username):
```
Only allow `admin`s to update events for now.

File: openlibrary/core/bookshelves_events.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
+from datetime import date, datetime
+from . import db
+
+
+class BookshelvesEvents:
```
Comment (jimchamp) [inline openlibrary/core/bookshelves_events.py:5]: We can easily make this a subclass of [CommonExtras](https://github.com/internetarchive/openlibrary/blob/e0e7a87d6cf69fb0b9cfb0ee7596657af247e056/openlibrary/core/db.py#L22) if it's deemed useful to do so.

File: openlibrary/plugins/upstream/checkins.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        """Validates POSTed check-in data."""
+        if not all(key in data for key in ('edition_olid', 'year', 'event_type')):
+            return False
+        if data['event_type'] not in BookshelvesEvents.EVENT_TYPES:
+            return False
+        return True
+
+
+class patron_check_ins(delegate.page):
+    path = r'/check-ins/people/([^/]+)'
```
Comment (cdrini) [inline openlibrary/plugins/upstream/checkins.py:78]: This Path pattern is common
https://openlibrary.org/people/ScarTissue/books/checkins

File: openlibrary/core/bookshelves_events.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+    TABLENAME = 'bookshelves_events'
+    EVENT_TYPES = {
+        'start': 1,
+        'update': 2,
+        'finish': 3,
+    }
```
Comment (cdrini) [inline openlibrary/core/bookshelves_events.py:12]: Python supports enums now:

```py
class BookshelfEvent:
    START = 1
    UPDATE = 2
    FINISH = 3
```

File: openlibrary/core/schema.sql [hunk 5]
```diff
@@ -52,6 +52,17 @@ INSERT INTO bookshelves (name, description) VALUES ('Want to Read', 'A list of b
 INSERT INTO bookshelves (name, description) VALUES ('Currently Reading', 'A list of books I am currently reading');
 INSERT INTO bookshelves (name, description) VALUES ('Already Read', 'A list of books I have finished reading');
 
+CREATE TABLE bookshelves_events (
+    id serial primary key,
+    username text not null,
+    work_id integer not null,
+    edition_id integer not null,
+    event_type integer not null,
```
Comment (cdrini) [inline openlibrary/core/schema.sql:60]: Is there a reason we didn't use postgres enums here?

See https://www.postgresql.org/docs/current/datatype-enum.html

File: openlibrary/core/schema.sql [hunk 6]
```diff
@@ -52,6 +52,17 @@ INSERT INTO bookshelves (name, description) VALUES ('Want to Read', 'A list of b
...
+    event_date text not null,
+    data json,
+    updated timestamp without time zone default (current_timestamp at time zone 'utc'),
+    created timestamp without time zone default (current_timestamp at time zone 'utc')
+);
```
Comment (cdrini) [inline openlibrary/core/schema.sql:65]: I'm not sure where we usually do this, but we will need to create indices over this table to make our queries performant.

File: openlibrary/plugins/upstream/checkins.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+        A request is invalid if it is:
+        a. Missing an 'id'
+        b. Does not have either 'year' or 'data'
+        """
+        if not 'id' in data:
+            return False
+        if not any(key in data for key in ('data', 'year')):
+            return False
+        return True
```
Comment (cdrini) [inline openlibrary/plugins/upstream/checkins.py:123]:
```suggestion
        return 'id' in data and 'data' in data and 'year' in data
```

File: openlibrary/core/bookshelves_events.py (hunk 2 above)
Comment (cdrini) [reply openlibrary/core/bookshelves_events.py:5]: Yeah we should, so that this table updates when work merges happen.

**Review by cdrini — COMMENTED** [review]
Looks good! Will test locally later. The manual enum is a little cumbersome and error-prone, if we could avoid it that would be great.

File: openlibrary/core/bookshelves_events.py (hunk 4 above)
Comment (cdrini) [reply openlibrary/core/bookshelves_events.py:12]: See https://docs.python.org/3/library/enum.html

File: openlibrary/core/schema.sql (hunk 5 above)
Comment (jimchamp) [reply openlibrary/core/schema.sql:60]: We use MySql, which handles enums differently, for testing.

File: openlibrary/core/schema.sql (hunk 6 above)
Comment (jimchamp) [reply openlibrary/core/schema.sql:65]: Let's have a better understanding of what we're querying before adding indices.  Can open an issue for this if you'd like.

**Comment by mekarpeles** [comment]
lgtm! looks like feedback has also all been addressed

**Pull request merged**